### PR TITLE
write tests

### PR DESCRIPTION
**Pull request by MykolaTurak**

### CODE REVIEW

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
you should write tests for all your handlers and services
Pay attention to the task description

File: src/test/java/core/basesyntax/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import core.basesyntax.infrastructure.db.FileReader;
+import core.basesyntax.infrastructure.db.FileReaderImpl;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+class MainTest {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/MainTest.java:11]: [Don't test the Main class](https://github.com/mate-academy/jv-fruit-shop-tests/blob/main/checklist.md#dont-test-the-main-class)

File: src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+class DataConverterImplTest {
+    @Test
+    void dataConverterTestStringOk() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java:12]: method called OK, but you expect an exception

File: src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        DataConverter dataConverter = new DataConverterImpl();
+        RuntimeException exception = assertThrows(RuntimeException.class,
+                () -> dataConverter.convertToTransaction(list));
+
+        assertEquals("Invalid number format: 'wrong'. Expected an integer value.",
+                exception.getMessage());
+    }
+
+    @Test
+    void dataConverterTestLessZeroOk() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java:24]: method called OK, but you expect an exception

File: src/test/java/core/basesyntax/infrastructure/db/FileReaderTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        };
+        List<String> expectedList = Arrays.asList(expectedStr);
+
+        FileReader fileReader = new FileReaderImpl();
+        List<String> actualList = fileReader.read(OPERATION_LIST_FILE_PATH);
+        assertEquals(expectedList, actualList);
+    }
+
+    @Test
+    void readInnitDataNotOk() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/db/FileReaderTest.java:35]: not clear why is not ok?

File: src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java (hunk 2 above)
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java:12]: Remember about the naming of test methods.

[same commment for all tests methods](https://github.com/mate-academy/jv-fruit-shop-tests/blob/main/checklist.md#remember-about-the-naming-of-test-methods)

File: src/test/java/core/basesyntax/service/FruitTransactionTest.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+class FruitTransactionTest {
+    @Test
+    void notFoundProduct() {
+        Storage.STORAGE.remove("banana");
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/FruitTransactionTest.java:18]: what exactly do you test here?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
add tests for all operations and services

File: src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        expected.add(new FruitTransaction(FruitTransaction.Operation.SUPPLY, "banana", 100));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.PURCHASE, "banana", 13));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.RETURN, "apple", 10));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.PURCHASE, "apple", 20));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.PURCHASE, "banana", 5));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.SUPPLY, "banana", 50));
+
+        FileReader reader = new FileReaderImpl();
+        List<String> read = reader.read("src/main/resources/operationslist.csv");
+        DataConverter converter = new DataConverterImpl();
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java:29]: `DataConverter converter` should be initialised in `@BeforeAll` method

File: src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import core.basesyntax.infrastructure.db.FileReader;
+import core.basesyntax.infrastructure.db.FileReaderImpl;
+import core.basesyntax.service.FruitTransaction;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+class DataConverterImplTest {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java:13]: for all classes
```suggestion
public class DataConverterImplTest {
```

File: src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Test
+    void convertToTransactionOk() {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java:16]: for all methods
```suggestion
    public void convertToTransactionOk() {
```

File: src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        expected.add(new FruitTransaction(FruitTransaction.Operation.BALANCE, "apple", 100));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.SUPPLY, "banana", 100));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.PURCHASE, "banana", 13));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.RETURN, "apple", 10));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.PURCHASE, "apple", 20));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.PURCHASE, "banana", 5));
+        expected.add(new FruitTransaction(FruitTransaction.Operation.SUPPLY, "banana", 50));
+
+        FileReader reader = new FileReaderImpl();
+        List<String> read = reader.read("src/main/resources/operationslist.csv");
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/DataConverterImplTest.java:28]: use constant for file path

File: src/test/java/core/basesyntax/infrastructure/db/FileReaderImplTest.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+class FileReaderImplTest {
+    @Test
+    void readInfoFromFileOk() {
+        FileWriter writer = new FileWriterImpl();
+        writer.write("banana,152\napple,90", "src/main/resources/database.csv");
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/db/FileReaderImplTest.java:14]: constant for file path

File: src/test/java/core/basesyntax/infrastructure/db/FileReaderImplTest.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+    }
+
+    @Test
+    void fileNotFoundExeptionOk() {
+        FileReader reader = new FileReaderImpl();
+        RuntimeException exception = assertThrows(RuntimeException.class,
+                () -> reader.read("/incorrect/path"));
+        assertEquals("Can't read from file", exception.getMessage());
+    }
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/db/FileReaderImplTest.java:33]: add test for empty file

File: src/test/java/core/basesyntax/infrastructure/db/FileWriterImplTest.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import org.junit.jupiter.api.Test;
+
+class FileWriterImplTest {
+    @Test
+    void writeIntoFileOk() {
+        String expectedMessage = "expected message";
+        List<String> expected = new ArrayList<>();
+        expected.add("expected message");
+        FileWriter fileWriter = new FileWriterImpl();
+        fileWriter.write("expected message", "src/main/resources/database.csv");
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/infrastructure/db/FileWriterImplTest.java:17]: constant for file path

File: src/test/java/core/basesyntax/service/FruitTransactionTest.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
+package core.basesyntax.service;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import org.junit.jupiter.api.Test;
+
+class FruitTransactionTest {
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/FruitTransactionTest.java:9]: redundnat test, don't need to check how Java methods works

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/OperationStrategyTest.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+public class OperationStrategyTest {
+
+    @Test
+    public void getOperationHandler() {
+        Map<FruitTransaction.Operation, OperationHandler> operationHandlers = new HashMap<>();
+        operationHandlers.put(FruitTransaction.Operation.BALANCE, new BalanceOperation());
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/OperationStrategyTest.java:16]: initialize Map in the method BeforeAll

File: src/test/java/core/basesyntax/service/OperationStrategyTest.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+import core.basesyntax.service.operations.BalanceOperation;
+import core.basesyntax.service.operations.OperationHandler;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+public class OperationStrategyTest {
+
+    @Test
+    public void getOperationHandler() {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/OperationStrategyTest.java:14]: add invalid test case

File: src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+public class ReportGeneratorImplTest {
+
+    @Test
+    public void getReportOk() {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java:18]: The same. Test ReportGenerator isolated. Add data to the Storage directly

File: src/test/java/core/basesyntax/service/ShopServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+import core.basesyntax.service.operations.PurchaseOperation;
+import core.basesyntax.service.operations.ReturnOperation;
+import core.basesyntax.service.operations.SupplyOperation;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+public class ShopServiceImplTest {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/ShopServiceImplTest.java:17]: the same comments

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/core/basesyntax/service/operations/SupplyOperation.java [hunk 18]
```diff
@@ -0,0 +1,14 @@
+package core.basesyntax.service.operations;
+
+import core.basesyntax.infrastructure.db.Storage;
+import core.basesyntax.service.FruitTransaction;
+
+public class SupplyOperation implements OperationHandler {
+    @Override
+    public void run(FruitTransaction fruitTransaction) {
+        this.isFruitAvailable(fruitTransaction);
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/service/operations/SupplyOperation.java]: Do you really need this check in this method?

File: src/test/java/core/basesyntax/infrastructure/db/FileReaderImplTest.java (hunk 11 above)
Comment (liliia-ponomarenko) [reply src/test/java/core/basesyntax/infrastructure/db/FileReaderImplTest.java:40]: not fixed

File: src/test/java/core/basesyntax/infrastructure/db/FileWriterImplTest.java [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+    void cleanUp() throws IOException {
+        Files.deleteIfExists(Path.of(TEST_PATH));
+    }
+
+    @Test
+    public void write_into_file_ok() {
+        String content = "expected message";
+        fileWriter.write(content, TEST_PATH);
+
+        List<String> actual = fileReader.read(TEST_PATH);
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/infrastructure/db/FileWriterImplTest.java]: Don't use FileReader for testing  FileWriter

File: src/test/java/core/basesyntax/service/operations/OperationHandlerTest.java [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+public class OperationHandlerTest {
+
+    @AfterEach
+    public void clearStorage() {
+        Storage.STORAGE.clear();
+    }
+
+    @Test
+    public void no_such_fruit_should_throw_exception() {
+        // Гарантуємо, що fruit "apple" відсутній
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/operations/OperationHandlerTest.java]: remove comments

File: src/test/java/core/basesyntax/service/operations/OperationHandlerTest.java [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+
+public class OperationHandlerTest {
+
+    @AfterEach
+    public void clearStorage() {
+        Storage.STORAGE.clear();
+    }
+
+    @Test
+    public void no_such_fruit_should_throw_exception() {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/operations/OperationHandlerTest.java]: use the same name pattern

File: src/test/java/core/basesyntax/service/operations/PurchaseOperationTest.java [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+    public void purchase_operation_ok() {
+        OperationHandler balanceOperation = new BalanceOperation();
+        balanceOperation.run(
+                new FruitTransaction(FruitTransaction.Operation.BALANCE, "banana", 20));
+
+        OperationHandler purchaseOperation = new PurchaseOperation();
+        purchaseOperation.run(
+                new FruitTransaction(FruitTransaction.Operation.PURCHASE, "banana", 10));
+
+        ReportGenerator reportGenerator = new ReportGeneratorImpl();
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/operations/PurchaseOperationTest.java]: the same. Don't use ReportGenerator

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Let's fix a few comments ;)

File: src/test/java/core/basesyntax/service/operations/ReturnOperationTest.java [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+    public void return_operation_ok() {
+        OperationHandler balanceOperation = new BalanceOperation();
+        balanceOperation.run(new FruitTransaction(
+                FruitTransaction.Operation.BALANCE, "banana", 20));
+
+        OperationHandler returnOperation = new ReturnOperation();
+        returnOperation.run(new FruitTransaction(
+                FruitTransaction.Operation.RETURN, "banana", 20));
+
+        ReportGenerator reportGenerator = new ReportGeneratorImpl();
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/operations/ReturnOperationTest.java]: compare data from storage, don't use ReportGenerator

File: src/test/java/core/basesyntax/service/operations/SupplyOperationTest.java [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+    @Test
+    public void supply_operation_ok() {
+        OperationHandler balanceOperation = new BalanceOperation();
+        balanceOperation.run(
+                new FruitTransaction(FruitTransaction.Operation.BALANCE, "banana", 20));
+
+        OperationHandler supplyOperation = new SupplyOperation();
+        supplyOperation.run(new FruitTransaction(FruitTransaction.Operation.SUPPLY, "banana", 20));
+
+        ReportGenerator reportGenerator = new ReportGeneratorImpl();
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/operations/SupplyOperationTest.java]: same here

File: src/test/java/core/basesyntax/service/operations/BalanceOperationTest.java [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+        Storage.STORAGE.clear();
+    }
+
+    @Test
+    public void balance_operation_ok() {
+        OperationHandler balanceOperation = new BalanceOperation();
+        balanceOperation.run(new FruitTransaction(
+                FruitTransaction.Operation.BALANCE, "banana", 20));
+
+        ReportGenerator reportGenerator = new ReportGeneratorImpl();
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/operations/BalanceOperationTest.java]: same here

File: src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+public class ReportGeneratorImplTest {
+
+    @AfterEach
+    void cleanUpStorage() {
+        Storage.STORAGE.clear();
+    }
+
+    @Test
+    public void get_report_ok() {
+        // Наповнюємо Storage напряму
```
Comment (Elena-Bruyako) [inline src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java]: remove comment

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

**Review by sarakhmen — APPROVED** [review]
lgtm
